File: ios/Classes/AudioEnginePlayerPlugin.swift
import Flutter
import UIKit

public class AudioEnginePlayerPlugin: NSObject, FlutterPlugin {
    private var audioEnginePlayer = AudioEnginePlayer()
    private var eventSink: FlutterEventSink?
    
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "audio_engine_player", binaryMessenger: registrar.messenger())
        let eventChannel = FlutterEventChannel(name: "audio_engine_player_events", binaryMessenger: registrar.messenger())
        let instance = AudioEnginePlayerPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
        eventChannel.setStreamHandler(instance)
    }
    
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "getPlatformVersion":
            result("iOS " + UIDevice.current.systemVersion)
        case "duration":
            let value = audioEnginePlayer.totalDuration
            result(value)
        case "volume":
            let value = audioEnginePlayer.volume
            result(value)
        case "isMute":
            let value = audioEnginePlayer.isMute
            result(value)
        case "enableFadeEffect":
            let value = audioEnginePlayer.enableFadeEffect
            result(value)
        case "speed":
            let value = audioEnginePlayer.speed
            result(value)
        case "currentPlayIndex":
            let currentPlayIndex = audioEnginePlayer.currentPlayIndex
            result(currentPlayIndex)
        case "isPlaying":
            let isPlaying = audioEnginePlayer.isPlaying
            result(isPlaying)
        case "ensureEngineRunning":
            audioEnginePlayer.ensureEngineRunning()
            result(nil)
        case "playWith":
            if let args = call.arguments as? [String: Any],
               let filePath = args["filePath"] as? String,
               let title = args["title"] as? String,
               let artist = args["artist"] as? String,
               let album = args["album"] as? String {
                let track = TrackModel(source: filePath, title: title, artist: artist, album: album, albumArt: nil)
                audioEnginePlayer.play(with: track)
                result(nil)
            } else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "filePath, title, artist, and album are required", details: nil))
            }
        case "seekTo":
            if let args = call.arguments as? [String: Any], let milliseconds = args["milliseconds"] as? Int {
                audioEnginePlayer.seekTo(milliseconds: milliseconds)
                result(nil)
            } else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "milliseconds is required", details: nil))
            }
        case "seekToIndex":
            if let args = call.arguments as? [String: Any], let index = args["index"] as? Int {
                audioEnginePlayer.seekToIndex(index);
                result(nil)
            } else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "milliseconds is required", details: nil))
            }
        case "playOrPause":
            audioEnginePlayer.playOrPause()
            result(nil)
        case "play":
            audioEnginePlayer.play()
            result(nil)
        case "pause":
            audioEnginePlayer.pause()
            result(nil)
        case "stop":
            audioEnginePlayer.stop()
            result(nil)
        case "setPlaylist":
            if let args = call.arguments as? [String: Any],
               let tracks = args["tracks"] as? [[String: Any]],
               let autoPlay = args["autoPlay"] as? Bool {
                let trackModels = tracks.compactMap { trackDict -> TrackModel? in
                    guard let source = trackDict["source"] as? String,
                          let title = trackDict["title"] as? String,
                          let artist = trackDict["artist"] as? String,
                          let album = trackDict["album"] as? String else {
                        return nil
                    }
                    return TrackModel(source: source, title: title, artist: artist, album: album, albumArt: nil)
                }
                audioEnginePlayer.setPlaylist(trackModels, autoPlay: autoPlay)
                result(nil)
            } else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "tracks and autoPlay are required", details: nil))
            }
        case "appendToPlaylist":
            if let args = call.arguments as? [String: Any],
               let source = args["source"] as? String,
               let title = args["title"] as? String,
               let artist = args["artist"] as? String,
               let album = args["album"] as? String,
               let autoPlay = args["autoPlay"] as? Bool {
                let track = TrackModel(source: source, title: title, artist: artist, album: album, albumArt: nil)
                audioEnginePlayer.appendToPlaylist(track, autoPlay: autoPlay)
                result(nil)
            } else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "source, title, artist, album, and autoPlay are required", details: nil))
            }
        case "removeFromPlaylist":
            if let args = call.arguments as? [String: Any], let index = args["index"] as? Int {
                audioEnginePlayer.removeFromPlaylist(index)
                result(nil)
            } else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "url and autoPlay are required", details: nil))
            }
        case "moveOnPlaylist":
            if let args = call.arguments as? [String: Any], let oldIndex = args["oldIndex"] as? Int, let newIndex = args["newIndex"] as? Int {
                audioEnginePlayer.moveOnPlaylist(oldIndex, newIndex);
                result(nil)
            } else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "url and autoPlay are required", details: nil))
            }
        case "playNext":
            audioEnginePlayer.playNext()
            result(nil)
        case "playPrevious":
            audioEnginePlayer.playPrevious()
            result(nil)
        case "setSpeed":
            if let args = call.arguments as? [String: Any], let value = args["speed"] as? Float {
                audioEnginePlayer.setSpeed(value)
                result(nil)
            } else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "speed is required", details: nil))
            }
        case "setVolume":
            if let args = call.arguments as? [String: Any], let volume = args["volume"] as? Float {
                audioEnginePlayer.setVolume(volume)
                result(nil)
            } else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "volume is required", details: nil))
            }
        case "setVolumeBoost":
            if let args = call.arguments as? [String: Any], let value = args["gain"] as? Float {
                audioEnginePlayer.setVolumeBoost(value)
                result(nil)
            } else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "volume is required", details: nil))
            }
        case "setMute":
            if let args = call.arguments as? [String: Any], let value = args["isMute"] as? Bool {
                audioEnginePlayer.setIsMute(value)
                result(nil)
            } else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "isMute is required", details: nil))
            }
        case "setEnableFadeEffect":
            if let args = call.arguments as? [String: Any], let value = args["enableFadeEffect"] as? Bool {
                audioEnginePlayer.setEnableFadeEffect(value)
                result(nil)
            } else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "isMute is required", details: nil))
            }
        case "setLoopMode":
            if let args = call.arguments as? [String: Any], let mode = args["mode"] as? Int {
                let loopMode: LoopMode
                switch mode {
                case 0:
                    loopMode = .all
                case 1:
                    loopMode = .single
                case 2:
                    loopMode = .shuffle
                default:
                    result(FlutterError(code: "INVALID_ARGUMENT", message: "Invalid loop mode", details: nil))
                    return
                }
                audioEnginePlayer.setLoopMode(loopMode)
                result(nil)
            } else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "mode is required", details: nil))
            }
        case "setBandGain":
            if let args = call.arguments as? [String: Any], let bandIndex = args["bandIndex"] as? Int, let gain = args["gain"] as? Float {
                audioEnginePlayer.setBandGain(bandIndex: bandIndex, gain: gain)
                result(nil)
            } else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "bandIndex and gain are required", details: nil))
            }
        case "setReverb":
            if let args = call.arguments as? [String: Any], let id = args["id"] as? Int, let wetDryMix = args["wetDryMix"] as? Float {
                audioEnginePlayer.setReverb(id: id, wetDryMix: wetDryMix)
                result(nil)
            } else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "id and wetDryMix are required", details: nil))
            }
        case "resetAll":
            audioEnginePlayer.resetAll()
            result(nil)
        case "clearCaches":
            audioEnginePlayer.clearCaches()
            result(nil)
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    private func setupCallbacks() {
        audioEnginePlayer.onPlayingStatusChanged = { [weak self] isPlaying in
            guard let self = self, let eventSink = self.eventSink else { return }
            eventSink(["event": "playingStatus", "isPlaying": isPlaying])
        }
        
        audioEnginePlayer.onPlayingIndexChanged = { [weak self] index in
            guard let self = self, let eventSink = self.eventSink else { return }
            eventSink(["event": "playingIndex", "currentIndex": index])
        }
        
        audioEnginePlayer.onPlaybackProgressUpdate = { [weak self] progress in
            guard let self = self, let eventSink = self.eventSink else { return }
            eventSink(["event": "playbackProgress", "progress": progress, "duration": self.audioEnginePlayer.totalDuration])
        }
        
        audioEnginePlayer.onPlayCompleted = { [weak self] in
            guard let self = self, let eventSink = self.eventSink else { return }
            eventSink(["event": "playCompleted", "isCompleted": true])
        }
    }
}

extension AudioEnginePlayerPlugin: FlutterStreamHandler {
    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        self.eventSink = events
        setupCallbacks()
        return nil
    }
    
    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        self.eventSink = nil
        return nil
    }
}

File: android/src/main/kotlin/com/ppsw/tech/audio_engine_player/AudioEnginePlayerPlugin.kt
package com.ppsw.tech.audio_engine_player

import androidx.annotation.NonNull
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

/** AudioEnginePlayerPlugin */
class AudioEnginePlayerPlugin: FlutterPlugin, MethodCallHandler {
  private lateinit var channel : MethodChannel
  private val audioEnginePlayer = AudioEnginePlayer()

  override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    channel = MethodChannel(flutterPluginBinding.binaryMessenger, "audio_engine_player")
    channel.setMethodCallHandler(this)
  }

  override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
    when (call.method) {
      "getPlatformVersion" -> result.success("Android ${android.os.Build.VERSION.RELEASE}")
      "getTotalDuration" -> result.success(audioEnginePlayer.totalDuration)
      "getVolume" -> result.success(audioEnginePlayer.volume)
      "play" -> {
        val filePath = call.argument<String>("filePath")
        if (filePath != null) {
          audioEnginePlayer.play(filePath)
          result.success(null)
        } else {
          result.error("INVALID_ARGUMENT", "filePath is required", null)
        }
      }
      "seekTo" -> {
        val milliseconds = call.argument<Int>("milliseconds")
        if (milliseconds != null) {
          audioEnginePlayer.seekTo(milliseconds)
          result.success(null)
        } else {
          result.error("INVALID_ARGUMENT", "milliseconds is required", null)
        }
      }
      "playOrPause" -> {
        audioEnginePlayer.playOrPause()
        result.success(null)
      }
      "stop" -> {
        audioEnginePlayer.stop()
        result.success(null)
      }
      "setPlaylist" -> {
        val urls = call.argument<List<String>>("urls")
        val autoPlay = call.argument<Boolean>("autoPlay") ?: false
        if (urls != null) {
          audioEnginePlayer.setPlaylist(urls, autoPlay)
          result.success(null)
        } else {
          result.error("INVALID_ARGUMENT", "urls and autoPlay are required", null)
        }
      }
      "appendToPlaylist" -> {
        val url = call.argument<String>("url")
        val autoPlay = call.argument<Boolean>("autoPlay") ?: false
        if (url != null) {
          audioEnginePlayer.appendToPlaylist(url, autoPlay)
          result.success(null)
        } else {
          result.error("INVALID_ARGUMENT", "url and autoPlay are required", null)
        }
      }
      "playNext" -> {
        audioEnginePlayer.playNext()
        result.success(null)
      }
      "playPrevious" -> {
        audioEnginePlayer.playPrevious()
        result.success(null)
      }
      "setVolume" -> {
        val volume = call.argument<Float>("volume")
        if (volume != null) {
          audioEnginePlayer.setVolume(volume)
          result.success(null)
        } else {
          result.error("INVALID_ARGUMENT", "volume is required", null)
        }
      }
      "setLoopMode" -> {
        val mode = call.argument<Int>("mode")
        if (mode != null) {
          val loopMode = when (mode) {
            0 -> LoopMode.ALL
            1 -> LoopMode.SINGLE
            2 -> LoopMode.SHUFFLE
            else -> {
              result.error("INVALID_ARGUMENT", "Invalid loop mode", null)
              return
            }
          }
          audioEnginePlayer.setLoopMode(loopMode)
          result.success(null)
        } else {
          result.error("INVALID_ARGUMENT", "mode is required", null)
        }
      }
      "setBandGain" -> {
        val bandIndex = call.argument<Int>("bandIndex")
        val gain = call.argument<Float>("gain")
        if (bandIndex != null && gain != null) {
          audioEnginePlayer.setBandGain(bandIndex, gain)
          result.success(null)
        } else {
          result.error("INVALID_ARGUMENT", "bandIndex and gain are required", null)
        }
      }
      "setReverb" -> {
        val id = call.argument<Int>("id")
        val wetDryMix = call.argument<Float>("wetDryMix")
        if (id != null && wetDryMix != null) {
          audioEnginePlayer.setReverb(id, wetDryMix)
          result.success(null)
        } else {
          result.error("INVALID_ARGUMENT", "id and wetDryMix are required", null)
        }
      }
      "resetAll" -> {
        audioEnginePlayer.resetAll()
        result.success(null)
      }
      "clearCaches" -> {
        audioEnginePlayer.clearCaches()
        result.success(null)
      }
      else -> result.notImplemented()
    }
  }

  override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
    channel.setMethodCallHandler(null)
  }
}